### PR TITLE
Make `derive(Trait)` suggestion more accurate

### PR DESCRIPTION
**Pull request by estebank**
Only suggest `derive(PartialEq)` when both LHS and RHS types are the same, otherwise the suggestion is not useful.

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 1]
```diff
@@ -2322,7 +2323,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             .iter()
             .map(|e| (e.obligation.predicate, None, Some(e.obligation.cause.clone())))
             .collect();
-        self.suggest_derive(err, &preds);
+        self.suggest_derive(err, &preds, suggest_derive);
     }
 
     pub fn suggest_derive(
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:2329]: I find it confusing that this method is called `suggest_derive` but apparently does more than suggesting `#[derive]`. It also adds the note *the trait[…] […] must be implemented*. Adding the flag `suggest_derive` to … `suggest_derive` makes it even more confusing. Could this method be renamed or even refactored into two methods potentially allowing us to get rid of the `bool` parameter altogether?

Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:2329]: I wanted to make this as minimal a diff as possible :)

`suggest_derive` *also* emits some extra notes (that I didn't want to lose). I can either rename the method and keep the argument, or split the logic in two (or three?) methods and call them only when needed.

Comment (fmease) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:2329]: I see :) If you can find a good name for it, go for a rename. You can do the refactor in this PR or or in a follow-up one, larger diffs are fine by me.

Comment (estebank) [reply compiler/rustc_hir_typeck/src/method/suggest.rs]: What do you think of the change?

Comment (fmease) [reply compiler/rustc_hir_typeck/src/method/suggest.rs]: Thanks, it's an improvement. Although, the way it's split into two methods doesn't feel right: `note_predicate_source_and_get_derives` still does two things, it calculates `traits` and `derives` and only adds a note for `traits` but never for `derive` and we do a bunch of work formatting `derives` just to throw it away if `!suggest_derive`.

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 2]
```diff
@@ -2322,18 +2323,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+    fn note_predicate_source_and_get_derives(
         &self,
         err: &mut Diagnostic,
         unsatisfied_predicates: &[(
             ty::Predicate<'tcx>,
             Option<ty::Predicate<'tcx>>,
             Option<ObligationCause<'tcx>>,
         )],
-    ) {
+    ) -> Vec<(String, Span, String)> {
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:2342]: Hmm, it's only in the error path but I'd love to skipp the entire `if let Some(diagnostic_name) = self.tcx.get_diagnostic_name(trait_pred.def_id())` check if `!suggest_derive` but then we'd need to introduce `suggested_derive: bool` to `note_pre_src_and_get_derives` :(

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 3]
```diff
@@ -2419,6 +2425,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             );
         }
 
+        derives_grouped
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:2428]: Could you move the code responsible for grouping of the derives from `note_pred_src` to `suggest_derive`?

**Review by fmease — APPROVED** [review]
r=me with the actionable subset of my most recent suggestions applied or if you find a better way to split those two functions then go for it and re-request a review :)

**Pull request merged**